### ISSUE TITLE
Better ressource distribution for ISRU (water specifically)

### ISSUE THREAD

**Issue by robot-aquila** [Kerbalism/Kerbalism#16]
1) It does not get in account resource abundance
2) Seems that harvesting does not work at all for unloaded (remote) vessel

TL;DR

I get a lot fun planning of missions. Thanks Kerbalism I got a wide field to work on.
In making an excel sheet to help me combine all together I have found some issues.
It does not look like as expected behaviour. And those are quite strange, I can't believe those are bugs because many people playing Kerbalism. But tests show it is it.

I started a fresh game to be sure.
Installed mods:

 * Kerbal Alarm Clock v3.9.1.0
 * Kerbalism 1.5.1
 * Module Manager 3.0.7
 * Community Resource Pack 0.10.0.0

I built two vessels (stock parts only): a vehicle with drill-o-matic jr wchich can drill near KSP continuously and a sattelite to switch to it to get driller unloaded. Water abundance near KSP is 25.68%.

TestCase1: Unloaded vessel test, max possible time warp
Time			Water on board
Y1, D3, 00:00:00	0.62
Y1, D4, 00:00:00	0.62
Y1, D5, 00:00:00	0.64
Y1, D6, 00:00:00	0.65
Note: it is literally visible when harvesting get faster after switching to vessel.

TestCase2: Loaded vessel test, max possible time warp

Time			Water	Change
Y1, D7, 00:00:00	11.36	+10.71
Y1, D8, 00:00:00	22.16	+10.80
Y1, D9, 00:00:00	32.94	+10.78
Y1, D10, 00:00:00	43.80	+10.86

TestCase3: Unloaded vessel test, realtime
Time			Water
Y1, D10, 01:00:00	43.81
Y1, D10, 02:00:00	43.82

Conclusions:
1) Time warp does not affect issue. It also in effect while playing realtime.
2) It is not ~25% abundance in effect because expected is 0.03/m*60*6*0.25=2.7/d but actual is 0.03/m*60*6=10.8/d yeld per day that is 100% not 25%.

Investigations:
1) Looking here src/Modules/Harvester.cs I do not see rate modification by abundance as well as no abundance applied while making an output (lines 94, 106).
2) The only place I can guess where is an issue is src/System/Kerbalism.cs FixedUpdate method. If local elapsed_s get zero then background update will give zero yeld. Possible affected on all production/consumption (but it is not checked).

[KerbalismMiningTest-save.zip](https://github.com/steamp0rt/Kerbalism/files/2014457/KerbalismMiningTest-save.zip)
[KSP.log](https://github.com/steamp0rt/Kerbalism/files/2014462/KSP.log)
[buildID64.txt](https://github.com/steamp0rt/Kerbalism/files/2014464/buildID64.txt)

**Comment by madman2003** [comment]
@robot-aquila Can you test https://github.com/steamp0rt/Kerbalism/pull/17 (the last commit contains a little green check mark next to it that will lead you to a jenkins instance where you can download the Kerbalism.zip file) if it fixes your troubles?

**Comment by robot-aquila** [comment]
Still testing. It gives wrong result in background + realtime.

**Comment by madman2003** [comment]
@robot-aquila Come up with proposals on what makes sense ;-) Ideally in the form of a pull request.
Why do you need so much water btw? I'm aware that when growing food we don't have a fully closed loop system yet. Because it's non-trivial to recover water from poop basically, and because current tech does not allow for 100% water recovery from urine. But are you using all the regeneration systems? (for example water recycler and humidity controller)

**Comment by madman2003** [comment]
@robot-aquila discussions are better suited to the discord server: https://discord.gg/pzgETC

**Comment by robot-aquila** [comment]
Still a draft but calculations/conclusions/proposals are done
[kerbalism-water-harvesting.zip](https://github.com/steamp0rt/Kerbalism/files/2056467/kerbalism-water-harvesting.zip)
I should check it few more times.
Any corrections are appreciated.

**Comment by PiezPiedPy** [comment]
I'll check it out when I get a bit more spare time 😉

**Comment by robot-aquila** [comment]
Oh. The document is big and I thought it might be uncomfortable. I guess it is better to place proposals directly here

Considering is Kerbalism v1.6.0.0
1. To keep ice water harvester output “as is” on 0.003/s
2. To start discussion on development for better water distribution model

**Comment by gotmachine** [comment]
Do we want our own resource abondance system ?
Keep in mind that things like that require support for every planet pack, and are likely to break compatibility with other mods.
Is there another mod that does it already ? 
Isn't there is some work on this aspect from Kopernicus ?

**Comment by gotmachine** [comment]
The [rational resource](https://forum.kerbalspaceprogram.com/index.php?/topic/184875-rational-resources-v091-aug-08-2019/) mod is doing interesting things in regard to resource distribution.

As far as I know, rational resources does 2 things :
1. New resource distribution for the stock (and some planet packs) bodies for CRP resources.
2. New ISRU modules / extractors for CRP resources, to replace the basic stock ore based system.

While Kerbalism could use 1. without problem, the patches implemented with 2. are redundant and incompatible with kerbalism own converters/extractors.
We could support 1. if those 2 parts of RR were distributed as 2 separate mods.

Maybe we should investigate how to support it instead of implementing something that will be (another) maintenance burden and another mod compatibility global issue.

**Comment by UsernamesAreAlwaysTaken** [comment]
2 is patcheable. that one is on me.

I'll get to it when i won't have so much shit on my plate.

**Comment by todi** [comment]
> We could support 1. if those 2 parts of RR were distributed as 2 separate mods.

Since version 1.0.0 Rational Resources is now indeed split into 2 mods: https://github.com/JadeOfMaar/RationalResources/releases/tag/1.0.0

**Comment by JadeOfMaar** [comment]
Hello, gents. Rational Resources was built with life support mods (particularly Kerbalism) in mind. And it's very easy (requiring little effort) to make any planet pack compatible but at the cost of rapidly inflating the MM cache due to the level of precision that Rational Resources boasts. 

I've just added documentation concerning the split mods [here](https://github.com/JadeOfMaar/RationalResources/wiki#mods) (thanks for drawing attention to that @todi ) and you can read about the templates and see how easy it is to setup a planet pack [here](https://github.com/JadeOfMaar/RationalResources/wiki/Templates).

@gotmachine Concerning the issue of maintenance headache, redundancy and incompatibility, I was personally told by, and I effectively agreed with the Kerbalism devs that once RR's ISRU chains were finalized, their figures would be ported over to Kerbalism's style of converter configs and mine would stop themselves from running when Kerbalism is present.
